modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala

```diff
@@ -505,118 +506,121 @@ private[smithy4s] class SchemaVisitorJCodec(
+          case a: DArray =>
+            out.writeArrayStart()
+            a.value match {
+              // short-circuiting on empty arrays to avoid the downcast to array of documents
+              // which has proven to be dangerous in Scala 3:
+              // https://github.com/disneystreaming/smithy4s/issues/1158
+              case x: ArraySeq[_] =>
+                if (x.isEmpty) ()
+                else {
+                  val xs = x.unsafeArray.asInstanceOf[Array[Document]]
+                  var i = 0
+                  while (i < xs.length) {
+                    encodeValue(xs(i), out)
+                    i += 1
+                  }
                 }
-              }
-            case xs =>
-              xs.foreach(encodeValue(_, out))
-          }
-          out.writeArrayEnd()
-        case o: DObject =>
-          out.writeObjectStart()
-          o.value.foreach { kv =>
-            out.writeKey(kv._1)
-            encodeValue(kv._2, out)
-          }
-          out.writeObjectEnd()
-        case _ => out.writeNull()
-      }
+              case xs =>
+                xs.foreach(encodeValue(_, out))
+            }
+            out.writeArrayEnd()
+          case o: DObject =>
+            out.writeObjectStart()
+            o.value.foreach { kv =>
+              out.writeKey(kv._1)
+              encodeValue(kv._2, out)
+            }
+            out.writeObjectEnd()
+          case _ => out.writeNull()
+        }
 
-      def decodeKey(in: JsonReader): Document =
-        in.decodeError("Cannot use JSON document as keys")
+        def decodeKey(in: JsonReader): Document =
+          in.decodeError("Cannot use JSON document as keys")
 
-      def encodeKey(x: Document, out: JsonWriter): Unit =
-        out.encodeError("Cannot use JSON documents as keys")
+        def encodeKey(x: Document, out: JsonWriter): Unit =
+          out.encodeError("Cannot use JSON documents as keys")
 
-      def expecting: String = "JSON document"
+        def expecting: String = "JSON document"
 
-      // Borrowed from: https://github.com/plokhotnyuk/jsoniter-scala/blob/e80d51019b39efacff9e695de97dce0c23ae9135/jsoniter-scala-benchmark/src/main/scala/io/circe/CirceJsoniter.scala
-      def decodeValue(cursor: Cursor, in: JsonReader): Document = {
-        val b = in.nextToken()
-        if (b == '"') {
-          in.rollbackToken()
-          new DString(in.readString(null))
-        } else if (b == 'f' || b == 't') {
-          in.rollbackToken()
-          new DBoolean(in.readBoolean())
-        } else if ((b >= '0' && b <= '9') || b == '-') {
-          in.rollbackToken()
-          new DNumber(in.readBigDecimal(null))
-        } else if (b == '[') {
-          new DArray({
-            if (in.isNextToken(']')) ArraySeq.empty[Document]
-            else
-              ArraySeq.unsafeWrapArray {
+        private val preserveKeyOrder =
+          preserveMapOrder || hints.has(PreserveKeyOrder)
```

Comment:
Left a comment on the other PR, but I'm wondering if we can get rid of `preserveMapOrder` in favor of using a transformation in the aws module to add the hint/trait rather than keeping this boolean around